internal/topo/operator/projectset_operator.go

```diff
@@ -23,43 +23,101 @@ import (
 
 type ProjectSetOperator struct {
 	SrfMapping map[string]struct{}
+	srfName    string
+
+	collectionRows []xsql.CloneAbleRow
 }
 
+// Apply implement UnOperation
+// ProjectSetOperator will extract the results from the set-returning-function into multi rows by aligning other columns
+// For tuple, ProjectSetOperator will do the following transform:
+// {"a":[1,2],"b":3} => {"a":1,"b":3},{"a":2,"b":3}
+// For Collection, ProjectSetOperator will do the following transform:
+// [{"a":[1,2],"b":3},{"a":[1,2],"b":4}] = > [{"a":"1","b":3},{"a":"2","b":3},{"a":"1","b":4},{"a":"2","b":4}]
 func (ps *ProjectSetOperator) Apply(_ api.StreamContext, data interface{}, _ *xsql.FunctionValuer, _ *xsql.AggregateFunctionValuer) interface{} {
 	// for now we only support 1 srf function in the field
 	srfName := ""
 	for k := range ps.SrfMapping {
 		srfName = k
 		break
 	}
+	ps.srfName = srfName
+	ps.collectionRows = make([]xsql.CloneAbleRow, 0)
+	r := make([]interface{}, 0)
 	switch input := data.(type) {
 	case error:
-		return []interface{}{input}
+		return input
 	case xsql.TupleRow:
-		aValue, ok := input.Value(srfName, "")
-		if !ok {
-			return fmt.Errorf("can't find the result from the %v function", srfName)
+		newTuples, err := ps.handleSRFRow(srfName, input)
+		if err != nil {
+			return err
+		}
+		for _, tuple := range newTuples {
+			r = append(r, tuple)
 		}
-		aValues, ok := aValue.([]interface{})
-		if !ok {
-			return fmt.Errorf("the result from the %v function should be array", srfName)
+		return r
+	case xsql.Collection:
+		err := input.Range(ps.handleSRFRowForCollection)
+		if err != nil {
+			return err
 		}
-		newData := make([]interface{}, 0)
-		for _, v := range aValues {
-			newTupleRow := input.Clone()
-			// clear original column value
-			newTupleRow.Del(srfName)
-			if mv, ok := v.(map[string]interface{}); ok {
-				for k, v := range mv {
-					newTupleRow.Set(k, v)
-				}
-			} else {
-				newTupleRow.Set(srfName, v)
+		switch ts := input.(type) {
+		case *xsql.JoinTuples:
+			newTuples := make([]*xsql.JoinTuple, 0)
+			for _, tuple := range ps.collectionRows {
+				newTuples = append(newTuples, tuple.(*xsql.JoinTuple))
+			}
+			ts.Content = newTuples
+		case *xsql.GroupedTuplesSet:
+			newTuples := make([]*xsql.GroupedTuples, 0)
```

Comment:
0 -》len(ps.collectionRows)

Reply:
Or make([]*xsql.GroupedTuples, 0, len(ps.collectionRows)) Or is it possible to have no new allocation/copy

Reply:
Other types also have this problem

Reply:
updated.